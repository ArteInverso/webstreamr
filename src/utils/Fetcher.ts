import CachePolicy from 'http-cache-semantics';
import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Context } from '../types';
import { BlockedError, NotFoundError } from '../error';
import { clearTimeout } from 'node:timers';

interface HttpCacheItem { policy: CachePolicy; status: number; statusText: string; body: string }

export class Fetcher {
  private readonly logger: winston.Logger;
  private readonly httpCache: TTLCache<string, HttpCacheItem>;

  constructor(logger: winston.Logger) {
    this.logger = logger;
    this.httpCache = new TTLCache();
  }

  readonly text = async (ctx: Context, url: URL, init?: RequestInit): Promise<string> => {
    return (await this.cachedFetch(ctx, url, init)).body;
  };

  readonly textPost = async (ctx: Context, url: URL, data: unknown, init?: RequestInit): Promise<string> => {
    return (await this.cachedFetch(ctx, url, { ...init, method: 'POST', body: JSON.stringify(data) })).body;
  };

  readonly head = async (ctx: Context, url: URL, init?: RequestInit): Promise<CachePolicy.Headers> => {
    return (await this.cachedFetch(ctx, url, { ...init, method: 'HEAD' })).policy.responseHeaders();
  };

  private readonly getInit = (ctx: Context, url: URL, init?: RequestInit): RequestInit => ({
    ...init,
    headers: {
      'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
      'Accept-Language': 'en',
      'Forwarded': `for=${ctx.ip}`,
      'Priority': 'u=0',
      'Referer': `${ctx.referer?.href ?? url.origin}`,
      'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/134.0.0.0 Safari/537.3',
      'X-Forwarded-For': ctx.ip,
      'X-Forwarded-Proto': url.protocol.slice(0, -1),
      'X-Real-IP': ctx.ip,
      ...init?.headers,
    },
  });

  private readonly handleHttpCacheItem = (httpCacheItem: HttpCacheItem): HttpCacheItem => {
    if (httpCacheItem.status && httpCacheItem.status >= 200 && httpCacheItem.status <= 299) {
      return httpCacheItem;
    }

    if (httpCacheItem.status === 404) {
      throw new NotFoundError();
    }

    const responseHeaders = httpCacheItem.policy.responseHeaders();

    if (httpCacheItem.policy.responseHeaders()['cf-mitigated'] === 'challenge') {
      throw new BlockedError('cloudflare_challenge');
    }

    if (httpCacheItem.status === 403) {
      throw new BlockedError('unknown');
    }

    throw new Error(`Fetcher error: ${httpCacheItem.status}: ${httpCacheItem.statusText}, response headers: ${JSON.stringify(responseHeaders)}`);
  };

  private readonly headersToObject = (headers: Headers): Record<string, string> => {
    const obj: Record<string, string> = {};

    headers.forEach((value, name) => {
      obj[name] = value;
    });

    return obj;
  };

  private readonly cachedFetch = async (ctx: Context, url: URL, init?: RequestInit): Promise<HttpCacheItem> => {
    const newInit = this.getInit(ctx, url, init);

    const request: CachePolicy.Request = { url: url.href, method: newInit.method ?? 'GET', headers: {} };

    let httpCacheItem: HttpCacheItem | undefined = this.httpCache.get(url.href);
    if (httpCacheItem) {
      this.logger.info(`Cached fetch ${request.method} ${url}`, ctx);
      return this.handleHttpCacheItem(httpCacheItem);
    }

    this.logger.info(`Fetch ${request.method} ${url}`, ctx);

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), 10000);

    let response;
    try {
      response = await fetch(url, { ...newInit, signal: controller.signal });
    } finally {
      clearTimeout(timer);
    }

    const body = await response.text();

    const policy = new CachePolicy(request, { status: response.status, headers: this.headersToObject(response.headers) }, { shared: false });

    httpCacheItem = { policy, status: response.status, statusText: response.statusText, body };

    let ttl;
    if (response.ok) {
      ttl = Math.max(policy.timeToLive(), 900000); // 15m at least
    } else {
      ttl = policy.timeToLive();
    }

    if (ttl > 0) {
      this.httpCache.set(url.href, httpCacheItem, { ttl });
    }

    return this.handleHttpCacheItem(httpCacheItem);
  };
}
