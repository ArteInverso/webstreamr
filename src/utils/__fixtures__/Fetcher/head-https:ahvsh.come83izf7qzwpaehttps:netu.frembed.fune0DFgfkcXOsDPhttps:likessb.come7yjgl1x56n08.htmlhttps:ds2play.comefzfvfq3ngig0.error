{"message":"socket hang up","code":"ECONNRESET"}