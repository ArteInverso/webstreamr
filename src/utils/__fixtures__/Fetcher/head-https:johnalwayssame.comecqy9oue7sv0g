{"server":"ddos-guard","connection":"keep-alive","keep-alive":"timeout=60","set-cookie":["__ddg8_=gmw8dCG6blfocUE2; Domain=.johnalwayssame.com; Path=/; Expires=Thu, 22-May-2025 16:46:44 GMT","__ddg10_=1747931204; Domain=.johnalwayssame.com; Path=/; Expires=Thu, 22-May-2025 16:46:44 GMT","__ddg9_=130.61.236.204; Domain=.johnalwayssame.com; Path=/; Expires=Thu, 22-May-2025 16:46:44 GMT","__ddg1_=BgkoUHamR633bMYVIX7f; Domain=.johnalwayssame.com; HttpOnly; Path=/; Expires=Fri, 22-May-2026 16:26:44 GMT","XSRF-TOKEN=eyJpdiI6InNsano3R3B5dW41QkpHNzZOd2lvcXc9PSIsInZhbHVlIjoiS3ZVTnZTRXJKc2F2OFh3MDFFWjhvZHpVOThVZnlZV2UzcXdxRVNOY2hWYTIvUUJCZnBtZXZiR295L0ozNXFGS0VmMlJ0VldpSmRKOE91Y2pjT0cxVVhXUDdvTG1HaDlFaldWaHIyTEJ2UysrMVJ4WGViUng5dGVEZmlqZHZqZ2EiLCJtYWMiOiJiYTJjY2U4ZGI2NjIwNTFhNDVhYzFiOWY5NzcxNTkxZWRlMGQxNmJhZTFiN2JiMTgxYjRhMjI5MmIyMzI1ZmM1IiwidGFnIjoiIn0%3D; expires=Thu, 22 May 2025 17:56:44 GMT; Max-Age=5400; path=/; secure; samesite=none; partitioned","voe_session=eyJpdiI6IjJPR1JsTHR6Y1VaSTJ1N08vQm9mbWc9PSIsInZhbHVlIjoiMnlLU244dVVyazVKUHBPWXdjdWg1bW1IY1U4Qkx6WmNSSHVXamRVMnpwOWRnRWdTT2VFTExYelN1ZU9mVTJJN0QvYy9BdGRwQ3ZPWjlGaS96TkpDaVY5S3VuQytoNUpZTHRoUy92OFcwWWgzWlZKbFlpTktXZThFTjZkK2xkbUIiLCJtYWMiOiI3YzU1OTZjZTg1OWJkOGYyN2VmMGNhY2Q5MzE0MzA4MTMwZjdiMDlkNmUxNzczZDkzNjI5MmQ3NWRmMTIzZTEwIiwidGFnIjoiIn0%3D; expires=Thu, 22 May 2025 17:56:44 GMT; Max-Age=5400; path=/; secure; httponly; samesite=none; partitioned"],"strict-transport-security":"max-age=2628000","content-security-policy":"upgrade-insecure-requests;","content-type":"text/html; charset=UTF-8","vary":"Accept-Encoding","cache-control":"no-cache, private","date":"Thu, 22 May 2025 16:26:44 GMT"}