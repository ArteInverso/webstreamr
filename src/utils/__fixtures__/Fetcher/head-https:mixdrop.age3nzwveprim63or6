{"date":"Thu, 22 May 2025 16:26:44 GMT","content-type":"text/html; charset=UTF-8","connection":"keep-alive","server":"cloudflare","expires":"Thu, 19 Nov 1981 08:52:00 GMT","cache-control":"no-store, no-cache, must-revalidate","pragma":"no-cache","nel":"{\"report_to\":\"cf-nel\",\"success_fraction\":0.0,\"max_age\":604800}","cf-cache-status":"DYNAMIC","report-to":"{\"group\":\"cf-nel\",\"max_age\":604800,\"endpoints\":[{\"url\":\"https://a.nel.cloudflare.com/report/v4?s=%2FXbIAsQRSGxA0wnx8arRq5DMZd1Y0Ta8fD4BgmvhZtMHPbNqh67JS2Z93kneJdciD6Gr4Olhbu0n6fF4cH7ZVOZDvNz05S1QwA8%3D\"}]}","set-cookie":["PHPSESSID=rhtto1gok9odeq9p12dfble0qo; SameSite=None; Secure; Path=/"],"cf-ray":"943dad4c9d2f364b-FRA","alt-svc":"h3=\":443\"; ma=86400"}