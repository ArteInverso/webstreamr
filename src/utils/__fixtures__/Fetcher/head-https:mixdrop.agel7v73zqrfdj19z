{"date":"Thu, 22 May 2025 16:26:44 GMT","content-type":"text/html; charset=UTF-8","connection":"keep-alive","server":"cloudflare","expires":"Thu, 19 Nov 1981 08:52:00 GMT","cache-control":"no-store, no-cache, must-revalidate","pragma":"no-cache","nel":"{\"report_to\":\"cf-nel\",\"success_fraction\":0.0,\"max_age\":604800}","cf-cache-status":"DYNAMIC","report-to":"{\"group\":\"cf-nel\",\"max_age\":604800,\"endpoints\":[{\"url\":\"https://a.nel.cloudflare.com/report/v4?s=CDKjYdtj2bDLLHYXqP%2BDiezHbUv3evtF3P6dLr8DBvwLnG8ZsDWDdw%2BDwSlwSklQIbIrh5g4yz9G7LSqAD3bvo4oR9tDhajV92I%3D\"}]}","set-cookie":["PHPSESSID=7j46e49jabt96ea6tu426cb88c; SameSite=None; Secure; Path=/"],"cf-ray":"943dad4c3acb9280-FRA","alt-svc":"h3=\":443\"; ma=86400"}