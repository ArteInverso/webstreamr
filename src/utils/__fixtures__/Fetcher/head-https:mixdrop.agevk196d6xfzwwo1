{"date":"Thu, 22 May 2025 16:26:44 GMT","content-type":"text/html; charset=UTF-8","connection":"keep-alive","server":"cloudflare","expires":"Thu, 19 Nov 1981 08:52:00 GMT","cache-control":"no-store, no-cache, must-revalidate","pragma":"no-cache","nel":"{\"report_to\":\"cf-nel\",\"success_fraction\":0.0,\"max_age\":604800}","cf-cache-status":"DYNAMIC","report-to":"{\"group\":\"cf-nel\",\"max_age\":604800,\"endpoints\":[{\"url\":\"https://a.nel.cloudflare.com/report/v4?s=vR%2BavpQfik4lUW5zXJ8NR2skrj2BTHirUBPSy7%2FbUQMUtKaER%2Fgb6sFItzEZtbQ9t6rTSahW2ZGHWeaRsXES%2F%2BJ7uDpMoaITmHE%3D\"}]}","set-cookie":["PHPSESSID=fr7a9ikrus94vpbotbar1qe3ka; SameSite=None; Secure; Path=/"],"cf-ray":"943dad4c8c6ba5fa-FRA","alt-svc":"h3=\":443\"; ma=86400"}