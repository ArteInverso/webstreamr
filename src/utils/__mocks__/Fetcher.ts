import fs from 'node:fs';
import Axios, { AxiosError, AxiosRequestConfig, AxiosResponseHeaders, RawAxiosResponseHeaders } from 'axios';
import slugify from 'slugify';
import { Context } from '../../types';

export class Fetcher {
  readonly text = async (_ctx: Context, url: URL, config?: AxiosRequestConfig): Promise<string> => {
    const path = `${__dirname}/../__fixtures__/Fetcher/${slugify(url.href)}`;

    return this.fixtureWrapper(path, async () => (await Axios.create().get(url.href, this.getConfig(config))).data);
  };

  readonly textPost = async (_ctx: Context, url: URL, data: unknown, config?: AxiosRequestConfig): Promise<string> => {
    const path = `${__dirname}/../__fixtures__/Fetcher/post-${slugify(url.href)}-${slugify(JSON.stringify(data))}`;

    return this.fixtureWrapper(path, async () => (await Axios.create().post(url.href, data, this.getConfig(config))).data);
  };

  readonly head = async (_ctx: Context, url: URL, config?: AxiosRequestConfig): Promise<RawAxiosResponseHeaders | AxiosResponseHeaders> => {
    const path = `${__dirname}/../__fixtures__/Fetcher/head-${slugify(url.href)}`;

    return JSON.parse(await this.fixtureWrapper(path, async () => JSON.stringify((await Axios.create().head(url.href, this.getConfig(config))).headers)));
  };

  private readonly fixtureWrapper = async (path: string, callable: () => Promise<string>): Promise<string> => {
    const errorPath = `${path}.error`;

    if (fs.existsSync(errorPath)) {
      const fixtureError = JSON.parse(fs.readFileSync(errorPath).toString());
      throw new AxiosError(fixtureError.message, fixtureError.code, undefined, undefined, fixtureError.response);
    } else if (fs.existsSync(path)) {
      return fs.readFileSync(path).toString();
    } else {
      let response;

      try {
        response = await callable();
      } catch (error) {
        if (error instanceof AxiosError) {
          const fixtureError = JSON.stringify({
            message: error.message,
            code: error.code,
            ...(error.response && {
              response: {
                status: error.response.status,
                statusText: error.response.statusText,
                data: error.response.data,
              },
            }),
          });
          fs.writeFileSync(errorPath, fixtureError);
        }

        throw error;
      }

      fs.writeFileSync(path, response);

      return response;
    }
  };

  private readonly getConfig = (config?: AxiosRequestConfig): AxiosRequestConfig => {
    return {
      responseType: 'text',
      ...config,
    };
  };
}
