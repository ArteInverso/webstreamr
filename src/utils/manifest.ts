import { flag } from 'country-emoji';
import { Handler } from '../handler';
import { Config, ManifestWithConfig } from '../types';

export const buildManifest = (handlers: Handler[], config: Config): ManifestWithConfig => {
  const manifest: ManifestWithConfig = {
    id: process.env['MANIFEST_ID'] || 'webstreamr',
    version: '0.17.0', // x-release-please-version
    name: process.env['MANIFEST_NAME'] || 'WebStreamr',
    description: 'Provides HTTP URLs from streaming websites.',
    resources: [
      'stream',
    ],
    types: [
      'movie',
      'series',
    ],
    catalogs: [],
    idPrefixes: ['tt'],
    behaviorHints: {
      p2p: false,
      configurable: true,
      configurationRequired: Object.keys(config).length === 0,
    },
    config: [],
  };

  const languageHandlers: Record<string, Handler[]> = {};
  handlers.forEach((handler) => {
    handler.languages.forEach(language => languageHandlers[language] = [...(languageHandlers[language] ?? []), handler]);
  });

  const sortedLanguageHandlers = Object.entries(languageHandlers)
    .sort(([languageA], [languageB]) => languageA.localeCompare(languageB));

  for (const [language, handlers] of sortedLanguageHandlers) {
    manifest.config.push({
      key: language,
      type: 'checkbox',
      title: `${flag(language)} (${handlers.map(handler => handler.label).join(', ')})`,
      ...(language in config && { default: 'checked' }),
    });
  }

  return manifest;
};
