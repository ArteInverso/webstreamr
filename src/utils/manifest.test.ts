import winston from 'winston';
import { buildManifest } from './manifest';
import { KinoKiste, MeineCloud, VerHdLink } from '../handler';
import { Fetcher } from './Fetcher';
import { ExtractorRegistry } from '../extractor';
jest.mock('../utils/Fetcher');

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
// @ts-expect-error No constructor args needed
const fetcher = new Fetcher();

describe('buildManifest', () => {
  test('has unchecked handler without a config', () => {
    const extractorRegistry = new ExtractorRegistry(logger, fetcher);
    const handlers = [
      new VerHdLink(fetcher, extractorRegistry),
      new KinoKiste(fetcher, extractorRegistry),
      new MeineCloud(fetcher, extractorRegistry),
    ];

    const manifest = buildManifest(handlers, {});

    expect(manifest.config).toMatchSnapshot();
  });

  test('has checked handler with appropriate config', () => {
    const extractorRegistry = new ExtractorRegistry(logger, fetcher);
    const handlers = [
      new VerHdLink(fetcher, extractorRegistry),
      new KinoKiste(fetcher, extractorRegistry),
      new MeineCloud(fetcher, extractorRegistry),
    ];
    const manifest = buildManifest(handlers, { de: 'on' });

    expect(manifest.config).toMatchSnapshot();
  });
});
