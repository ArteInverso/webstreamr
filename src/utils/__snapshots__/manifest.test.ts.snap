// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildManifest has checked handler with appropriate config 1`] = `
[
  {
    "default": "checked",
    "key": "de",
    "title": "German ðŸ‡©ðŸ‡ª (KinoKiste, MeineCloud)",
    "type": "checkbox",
  },
  {
    "key": "es",
    "title": "Castilian Spanish ðŸ‡ªðŸ‡¸ (VerHdLink)",
    "type": "checkbox",
  },
  {
    "key": "mx",
    "title": "Latin American Spanish ðŸ‡²ðŸ‡½ (VerHdLink)",
    "type": "checkbox",
  },
  {
    "key": "excludeExternalUrls",
    "title": "Exclude external URLs from results",
    "type": "checkbox",
  },
]
`;

exports[`buildManifest has unchecked handler without a config 1`] = `
[
  {
    "key": "de",
    "title": "German ðŸ‡©ðŸ‡ª (KinoKiste, MeineCloud)",
    "type": "checkbox",
  },
  {
    "key": "es",
    "title": "Castilian Spanish ðŸ‡ªðŸ‡¸ (VerHdLink)",
    "type": "checkbox",
  },
  {
    "key": "mx",
    "title": "Latin American Spanish ðŸ‡²ðŸ‡½ (VerHdLink)",
    "type": "checkbox",
  },
  {
    "key": "excludeExternalUrls",
    "title": "Exclude external URLs from results",
    "type": "checkbox",
  },
]
`;
